### PR TITLE
Flaky test reporting(part #4) Github issues tracking for flaky tests

### PR DESCRIPTION
**Pull request by chaodaiG**
Creating/updating/closing flaky tests related issues, based on test results collected from most recent continuous flows and data structure defined in [PR465](https://github.com/knative/test-infra/pull/465).

Each flaky test is tracked by one issue, and if there are too many tests found flaky there will be a bulk issue created instead of flooding issues pool. Issue identity is embedded in a reserved part of issue body instead of title, so developers have enough freedom of making adjustment to title and other parts of issue body. History of test results are tracked by prepend latest result to the first comment, and all old histories are hidden by a collapsible Markdown template. Other than text description, test results history is also represented by a unicode graph similar to TestGrid.

Part of: #132

### CODE REVIEW

**Comment by chaodaiG** [comment]
Example issues from fake data:
[Constantly flaky test](https://github.com/chaodaiG/test-github-api/issues/96)
[Fixed flaky test](https://github.com/chaodaiG/test-github-api/issues/98)
[Pseudo fixed test(passed one day, then failed next day)](https://github.com/chaodaiG/test-github-api/issues/97)
[bulk issue](https://github.com/chaodaiG/test-github-api/issues/106)

**Comment by chaodaiG** [comment]
Example issues from fake data:
[Constantly flaky test](https://github.com/chaodaiG/test-github-api/issues/96)
[Fixed flaky test](https://github.com/chaodaiG/test-github-api/issues/98)
[Pseudo fixed test(passed one day, then failed next day)](https://github.com/chaodaiG/test-github-api/issues/97)
[bulk issue](https://github.com/chaodaiG/test-github-api/issues/106)

File: tools/flaky-test-reporter/run.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// run.go controls how to run functions that needs dryrun support
+
+package main
```
Comment (srinivashegde86) [inline tools/flaky-test-reporter/run.go]: Update package name from main for all these files.

Comment (chaodaiG) [reply tools/flaky-test-reporter/run.go]: had an offline discussion with @srinivashegde86 , this certainly is a good idea, will do the package separation in subsequent PRs

File: tools/flaky-test-reporter/github_issue.go [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+		hasClosed := false
+		for _, fi := range v {
+			switch fi.issue.GetState() {
+				case string(ghutil.IssueOpenState):
+					hasOpen = true
+				case string(ghutil.IssueCloseState):
+					hasClosed = true
+			}
+		}
+		fmt.Sprintln(k, hasOpen, hasClosed)
```
Comment (adrcunha) [inline tools/flaky-test-reporter/github_issue.go]: Debug? If so, remove.

File: tools/flaky-test-reporter/github_issue.go [hunk 3]
```diff
@@ -0,0 +1,460 @@
+/*
```
Comment (adrcunha) [inline tools/flaky-test-reporter/github_issue.go]: As a general comment, I miss comments identifying large blocks of code. e.g., something like "// close open issue if the test passed" on line 195.

File: tools/flaky-test-reporter/github_issue.go [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+	}
+	if flakyRate > threshold {
+		log.Printf("flaky rate above '%f', creating a single issue", threshold)
+		identity := fmt.Sprintf("%.2f%% tests failed in repo %s on %s",
+			flakyRate * 100, rd.Config.Repo, time.Unix(*rd.LastBuildStartTime, 0).String())
+		if _, ok := flakyIssuesMap[identity]; ok {
+			log.Printf("issue already exist, skip creating")
+			return nil, nil
+		}
+		title := fmt.Sprintf("[flaky]%s", identity)
```
Comment (adrcunha) [inline tools/flaky-test-reporter/github_issue.go]:
```suggestion
		title := fmt.Sprintf("[flaky] %s", identity)
```

File: tools/flaky-test-reporter/github_issue.go [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+		comment := fmt.Sprintf("Bulk issue trakcing: %s\n%s", identity, fmt.Sprintf(testIdentifierPattern, identity))
```
Comment (adrcunha) [inline tools/flaky-test-reporter/github_issue.go]:
```suggestion
		comment := fmt.Sprintf("Bulk issue tracking: %s\n%s", identity, fmt.Sprintf(testIdentifierPattern, identity))
```

File: tools/flaky-test-reporter/github_issue.go [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+				return nil != issuesMap[k][i].issue.CreatedAt && 
+				( nil == issuesMap[k][j].issue.CreatedAt || issuesMap[k][i].issue.CreatedAt.Before(*issuesMap[k][j].issue.CreatedAt))
+			})
+			issuesMap[k] = []*flakyIssue{issuesMap[k][0]}
+		}
+	}
+	return issuesMap, err
+}
+
+func processGithubIssueForRepo(rd *RepoData, flakyIssuesMap map[string][]*flakyIssue, dryrun *bool) ([]string, error) {
```
Comment (adrcunha) [inline tools/flaky-test-reporter/github_issue.go]: Would be nice to document this function, specially the return values.

File: tools/flaky-test-reporter/github_issue.go [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+				message := fmt.Sprintf("Updating issue '%s' for '%s'", *existIssue.issue.URL, *existIssue.identity)
+				log.Println(message)
+				messages = append(messages, message)
+				if err := updateIssue(existIssue, comment, ts, dryrun); nil != err {
+					log.Println(err)
+					errs = append(errs, err)
+				}
+			}
+		} else if ts.isFlaky() {
+			title := fmt.Sprintf("[flaky]'%s'", identity)
```
Comment (adrcunha) [inline tools/flaky-test-reporter/github_issue.go]:
```suggestion
			title := fmt.Sprintf("[flaky] %s", identity)
```

File: tools/flaky-test-reporter/github_issue.go [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+	
+	oldHistory := regexp.MustCompile(reHistory).FindStringSubmatch(comment)
+	if len(oldHistory) >= 2 {
+		oldHistoryEntries := strings.Split(oldHistory[1], "\n")
+		if len(oldHistoryEntries) > 9 {
+			oldHistoryEntries = oldHistoryEntries[:9]
+		}
+		res = fmt.Sprintf("\n%s%s", currentUnicode, strings.Join(oldHistoryEntries, "\n"))
+	} else {
+		res = fmt.Sprintf("\n%s\n", currentUnicode)
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: The `\n` at line 149 is needed for the history to be displayed correctly, which is not needed at line 147, so it's not really avoidable.

File: tools/flaky-test-reporter/github_issue.go [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+	failedUnicode       = "&#10006;" // Cross unicode
+	skippedUnicode      = "&#9723;" // Open square unicode
+
+	// collapseTemplate is for creating collapsible format in comment, so that only latest result is expanded.
+	// Github markdown supports collapse, strings between <summary></summary> is the title,
+	// contents between <p></p> is collapsible body, body is hidden unless the title is clicked.
+	collapseTemplate    = `
+<details>
+	<summary>%s</summary><p>
+
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: yes, it's crucial for collapsible, comment added

File: tools/flaky-test-reporter/github_issue.go [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+// createCommentForTest summarizes latest status of current test case,
+// and creates text to be added to issue comment
+func createCommentForTest(rd *RepoData, testFullName string) string {
+	ts := rd.TestStats[testFullName]
+	totalCount := len(ts.Passed) + len(ts.Skipped) + len(ts.Failed)
+	lastBuildStartTimeStr := time.Unix(*rd.LastBuildStartTime, 0).String()
+	content := fmt.Sprintf(`%s
+Time for last build: %s
+Failed %d times out of %d runs`,
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: done

File: tools/flaky-test-reporter/github_issue.go [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+	ts := rd.TestStats[testFullName]
+	totalCount := len(ts.Passed) + len(ts.Skipped) + len(ts.Failed)
+	lastBuildStartTimeStr := time.Unix(*rd.LastBuildStartTime, 0).String()
+	content := fmt.Sprintf(`%s
+Time for last build: %s
+Failed %d times out of %d runs`, 
+		fmt.Sprintf(latestStatusPattern, ts.getTestStatus()),
+		lastBuildStartTimeStr, len(ts.Failed), totalCount)
+	if len(ts.Failed) > 0 {
+		content += ". Failed runs: "
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: Good catch, fixed

File: tools/flaky-test-reporter/github_issue.go [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+	return fmt.Sprintf(historyPattern, res)
+}
+
+// prependComment hides old comment into a collapsible, and prepend
+// new commment on top
+func prependComment(oldComment, newComment string) string {
+	if "" != oldComment {
+		oldComment = fmt.Sprintf(collapseTemplate, "Click to see older results", oldComment)
+	}
+	return fmt.Sprintf(`%s
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: done

File: tools/flaky-test-reporter/github_issue.go [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+// reopens the issue if test becomes flaky while issue is closed.
+func updateIssue(fi *flakyIssue, newComment string, ts *TestStat, dryrun *bool) error {
+	issue := fi.issue
+	passedLastTime := false
+	latestStatus := regexp.MustCompile(reLastestStatus).FindStringSubmatch(fi.comment.GetBody())
+	if len(latestStatus) >= 2{
+		switch latestStatus[1] {
+			case passedStatus:
+				passedLastTime = true
+			case flakyStatus, failedStatus, lackDataStutus:
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: fixed

File: tools/flaky-test-reporter/github_issue.go (hunk 2 above)
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: yeah it is, removed

File: tools/flaky-test-reporter/github_issue.go (hunk 3 above)
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: good idea, added

File: tools/flaky-test-reporter/github_issue.go (hunk 6 above)
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: done

File: tools/flaky-test-reporter/github_issue.go [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+			default:
+				statusUnicode = skippedUnicode
+		}
+		currentUnicode += fmt.Sprintf(" [%s](%s)", statusUnicode, url)
+	}
+	
+	oldHistory := regexp.MustCompile(reHistory).FindStringSubmatch(comment)
+	if len(oldHistory) >= 2 {
+		oldHistoryEntries := strings.Split(oldHistory[1], "\n")
+		if len(oldHistoryEntries) > 9 {
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: Made is a const, comment added

**Comment by chaodaiG** [comment]
> Since this is not a stand-alone tool, but a tool that interacts with GitHub, I miss having unit tests for it.

I will investigate how to add unit tests for this tool, and submit a separate PR, since this one is already large

File: tools/flaky-test-reporter/github_issue.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (adrcunha) [reply tools/flaky-test-reporter/github_issue.go]: `res` always starts with `currentUnicode`. The difference will be if you append `oldHistoryEntries` or just a linefeed.

**Comment by adrcunha** [comment]
One last nit, but I won't hold on it.
/lgtm
/hold

File: tools/flaky-test-reporter/github_issue.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chaodaiG) [reply tools/flaky-test-reporter/github_issue.go]: gotcha, updated

**Comment by chaodaiG** [comment]
> One last nit, but I won't hold on it.
> /lgtm
> /hold

the comment makes sense to me, pushed with updates. needs lgtm again @adrcunha

**Pull request merged**